File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/PaginationTest.java
package com.kinvey.androidTest.store;

import android.content.Context;
import android.os.Message;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.test.RenamingDelegatingContext;
import android.test.suitebuilder.annotation.SmallTest;

import com.kinvey.android.Client;
import com.kinvey.android.model.User;
import com.kinvey.android.store.DataStore;
import com.kinvey.android.store.UserStore;
import com.kinvey.androidTest.LooperThread;
import com.kinvey.androidTest.TestManager;
import com.kinvey.androidTest.callback.CustomKinveyPullCallback;
import com.kinvey.androidTest.callback.CustomKinveySyncCallback;
import com.kinvey.androidTest.model.Person;
import com.kinvey.java.Constants;
import com.kinvey.java.core.KinveyClientCallback;
import com.kinvey.java.model.KinveyPullResponse;
import com.kinvey.java.store.StoreType;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.util.ArrayList;
import java.util.concurrent.CountDownLatch;

import static com.kinvey.androidTest.TestManager.PASSWORD;
import static com.kinvey.androidTest.TestManager.TEST_USERNAME;
import static com.kinvey.androidTest.TestManager.TEST_USERNAME_2;
import static com.kinvey.androidTest.TestManager.USERNAME;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

@RunWith(AndroidJUnit4.class)
@SmallTest
public class PaginationTest {

    private Client client;
    private TestManager<Person> testManager;

    @Before
    public void setUp() throws InterruptedException {
        Context mMockContext = new RenamingDelegatingContext(InstrumentationRegistry.getInstrumentation().getTargetContext(), "test_");
        client = new Client.Builder(mMockContext).build();
        testManager = new TestManager<>();
        testManager.login(USERNAME, PASSWORD, client);

    }

    @After
    public void tearDown() {
        client.performLockDown();
        if (client.getKinveyHandlerThread() != null) {
            try {
                client.stopKinveyHandlerThread();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }


    @Test
    public void testPagedPull() throws InterruptedException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.CACHE, client);
        client.getSyncManager().clear(Person.COLLECTION);

        testManager.cleanBackendDataStore(store);

        // Arrange
        Person curly = new Person("Curly");
        Person larry = new Person("Larry");
        Person moe = new Person("Moe");
        testManager.save(store,curly);
        testManager.save(store,larry);
        testManager.save(store,moe);
        long cacheSizeBefore = testManager.getCacheSize(StoreType.CACHE, client);
        assertTrue(cacheSizeBefore == 3);
        client.getCacheManager().getCache(Person.COLLECTION, Person.class, StoreType.CACHE.ttl).clear();
        long cacheSizeBetween = testManager.getCacheSize(StoreType.CACHE, client);
        assertTrue(cacheSizeBetween == 0);

        // Act
        CustomKinveyPullCallback pullCallback = testManager.pullCustom(store, null, 2);

        // Assert
        assertNull(pullCallback.getError());
        assertNotNull(pullCallback.getResult());
        assertTrue(pullCallback.getResult().getCount() == 3);
        assertTrue(pullCallback.getResult().getCount() == testManager.getCacheSize(StoreType.CACHE,client));
    }

    @Test
    public void testPagedPullBlocking() throws InterruptedException, IOException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.CACHE, client);
        client.getSyncManager().clear(Person.COLLECTION);

        testManager.cleanBackendDataStore(store);

        // Arrange
        ArrayList<Person> persons = new ArrayList<>();
        Person alvin = new Person("Alvin");
        Person simon = new Person("Simon");
        Person theodore = new Person("Theodore");
        testManager.save(store,alvin);
        testManager.save(store,simon);
        testManager.save(store,theodore);
        long cacheSizeBefore = testManager.getCacheSize(StoreType.CACHE, client);
        assertTrue(cacheSizeBefore == 3);
        client.getCacheManager().getCache(Person.COLLECTION, Person.class, StoreType.CACHE.ttl).clear();
        long cacheSizeBetween = testManager.getCacheSize(StoreType.CACHE, client);
        assertTrue(cacheSizeBetween == 0);

        // Act
        KinveyPullResponse pullResponse = store.pullBlocking(null, 2);
        assertEquals(3, pullResponse.getCount());
        assertTrue(pullResponse.getCount() == testManager.getCacheSize(StoreType.CACHE, client));
    }


    @Test
    public void testSyncBlockingPaged() throws InterruptedException, IOException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        testManager.cleanBackend(store, StoreType.SYNC);
        client.getSyncManager().clear(Person.COLLECTION);
        Person person = new Person(TEST_USERNAME);
        testManager.save(store,person);
        assertTrue(client.getSyncManager().getCount(Person.COLLECTION) == 1);
        store.pushBlocking();
        person = new Person(TEST_USERNAME_2);
        testManager.save(store,person);
        store.syncBlocking(client.query(), 1);
        assertTrue(client.getSyncManager().getCount(Person.COLLECTION) == 0);
        assertTrue(testManager.find(store, client.query()).getResult().getResult().size() == 2);
    }

    @Test
    public void testSyncPaged() throws InterruptedException, IOException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        testManager.cleanBackend(store, StoreType.SYNC);
        client.getSyncManager().clear(Person.COLLECTION);
        testManager.createPersons(store, 10);
        assertTrue(client.getSyncManager().getCount(Person.COLLECTION) == 10);
        CustomKinveySyncCallback syncCallback = testManager.sync(store, null, 1);
        assertNull(syncCallback.getError());
        assertNotNull(syncCallback.getKinveyPushResponse().getSuccessCount());
        assertEquals(10, syncCallback.getKinveyPushResponse().getSuccessCount());
        assertNotNull(syncCallback.getResult());
        assertEquals(0, syncCallback.getResult().getListOfExceptions().size());
        assertEquals(10, syncCallback.getResult().getCount());
        assertTrue(client.getSyncManager().getCount(Person.COLLECTION) == 0);
        assertEquals(10, store.find().getResult().size());
    }

    @Test
    public void testSyncPagedWithQuery() throws InterruptedException, IOException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        testManager.cleanBackend(store, StoreType.SYNC);
        client.getSyncManager().clear(Person.COLLECTION);
        testManager.createPersons(store, 10);
        assertTrue(client.getSyncManager().getCount(Person.COLLECTION) == 10);
        CustomKinveySyncCallback syncCallback = testManager.sync(store, client.query().equals("username", TEST_USERNAME + 0).or(client.query().equals("username", TEST_USERNAME + 1)), 1);
        assertNull(syncCallback.getError());
        assertNotNull(syncCallback.getKinveyPushResponse().getSuccessCount());
        assertEquals(10, syncCallback.getKinveyPushResponse().getSuccessCount());
        assertNotNull(syncCallback.getResult());
        assertEquals(0, syncCallback.getResult().getListOfExceptions().size());
        assertEquals(2, syncCallback.getResult().getCount());
        assertTrue(client.getSyncManager().getCount(Person.COLLECTION) == 0);
        assertEquals(10, store.find().getResult().size());
    }


    @Test
    public void testPagedPullNotCorrectItem() throws InterruptedException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION_WITH_EXCEPTION, Person.class, StoreType.SYNC, client);
        CustomKinveyPullCallback pullCallback = testManager.pullCustom(store, null, 2);
        assertTrue(pullCallback.getResult().getListOfExceptions().size() == 1);
        assertTrue(pullCallback.getResult().getCount() == 4);
        testManager.cleanBackend(store, StoreType.SYNC);
    }


    @Test
    public void testPullBlockingWithAutoPagination() throws InterruptedException, IOException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        client.getSyncManager().clear(Person.COLLECTION);
        testManager.cleanBackend(store, StoreType.SYNC);

        testManager.createPersons(store, 5);
        testManager.push(store);
        client.getCacheManager().getCache(Person.COLLECTION, Person.class, StoreType.SYNC.ttl).clear();

        assertEquals(5, store.pullBlocking(client.query(), true).getCount());
        assertEquals(5, testManager.getCacheSize(StoreType.SYNC, client));
    }

    @Test
    public void testPullBlockingWithoutAutoPagination() throws InterruptedException, IOException {
        DataStore<Person> store = DataStore.collection(Person.COLLECTION, Person.class, StoreType.SYNC, client);
        client.getSyncManager().clear(Person.COLLECTION);

        testManager.cleanBackend(store, StoreType.SYNC);
        testManager.createPersons(store, 5);
        testManager.push(store);
        client.getCacheManager().getCache(Person.COLLECTION, Person.class, StoreType.SYNC.ttl).clear();

        assertEquals(5, store.pullBlocking(client.query(), false).getCount());
        assertEquals(5, testManager.getCacheSize(StoreType.SYNC, client));
    }


}

File: android-lib/src/androidTest/java/com/kinvey/androidTest/store/datastore/PaginationTest.kt
package com.kinvey.androidTest.store.datastore

import androidx.test.filters.SmallTest
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.runner.AndroidJUnit4
import com.kinvey.android.Client
import com.kinvey.android.Client.Builder
import com.kinvey.android.Client.Companion.kinveyHandlerThread
import com.kinvey.android.model.User
import com.kinvey.android.store.DataStore.Companion.collection
import com.kinvey.androidTest.TestManager
import com.kinvey.androidTest.TestManager.*
import com.kinvey.androidTest.TestManager.Companion.PASSWORD
import com.kinvey.androidTest.TestManager.Companion.TEST_USERNAME
import com.kinvey.androidTest.TestManager.Companion.TEST_USERNAME_2
import com.kinvey.androidTest.TestManager.Companion.USERNAME
import com.kinvey.androidTest.callback.CustomKinveyPullCallback
import com.kinvey.androidTest.model.Person
import com.kinvey.java.AbstractClient
import com.kinvey.java.core.AbstractKinveyClient
import com.kinvey.java.store.StoreType
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.lang.reflect.Field
import java.util.*

@RunWith(AndroidJUnit4::class)
@SmallTest
class PaginationTest {

    private var client: Client<*>? = null
    private var testManager: TestManager<Person>? = null

    @Before
    @Throws(InterruptedException::class)
    fun setUp() {
        val mMockContext = InstrumentationRegistry.getInstrumentation().targetContext
        client = Builder<User>(mMockContext).build()
        testManager = TestManager()
        testManager?.login(USERNAME, PASSWORD, client)
    }

    @After
    fun tearDown() {
        client?.performLockDown()
        if (kinveyHandlerThread != null) {
            try {
                client?.stopKinveyHandlerThread()
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
            }
        }
    }

    @Test
    @Throws(InterruptedException::class)
    fun testPagedPull() {
        val store = collection(COLLECTION, Person::class.java, StoreType.CACHE, client)
        client?.syncManager?.clear(COLLECTION)
        testManager?.cleanBackendDataStore(store)

        // Arrange
        val curly = Person("Curly")
        val larry = Person("Larry")
        val moe = Person("Moe")
        testManager?.save(store, curly)
        testManager?.save(store, larry)
        testManager?.save(store, moe)

        val cacheSizeBefore = testManager?.getCacheSize(StoreType.CACHE, client)
        assertTrue(cacheSizeBefore == 3L)
        client?.cacheManager?.getCache(COLLECTION, Person::class.java, StoreType.CACHE.ttl)?.clear()

        val cacheSizeBetween = testManager?.getCacheSize(StoreType.CACHE, client)
        assertTrue(cacheSizeBetween == 0L)

        // Act
        val pullCallback = testManager?.pullCustom(store, null, 2)

        // Assert
        assertNull(pullCallback?.error)
        assertNotNull(pullCallback?.result)
        assertTrue(pullCallback?.result?.count == 3)
        assertTrue(pullCallback?.result?.count?.toLong() == testManager?.getCacheSize(StoreType.CACHE, client))
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun syncErrorPaginationAsync() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.syncManager?.clear(COLLECTION)

        testManager?.createPersons(store, 10)
        assertTrue(client?.syncManager?.getCount(COLLECTION) == 10L)
        mockInvalidConnection()
        val syncCallback = testManager?.sync(store, null, false)
        cancelMockInvalidConnection()
        assertNull(syncCallback?.kinveyPushResponse)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSyncPagedError() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.syncManager?.clear(COLLECTION)

        testManager?.createPersons(store, 10)
        assertTrue(client?.syncManager?.getCount(COLLECTION) == 10L)
        mockInvalidConnection()
        val syncCallback = testManager?.sync(store, client?.query()?.equals("username", TestManager.TEST_USERNAME + 0)
                ?.or(client?.query()?.equals("username", TestManager.TEST_USERNAME + 1)), 1)
        cancelMockInvalidConnection()
        assertNull(syncCallback?.kinveyPushResponse)
    }

    fun mockInvalidConnection() {
        var field: Field? = null
        try {
            field = AbstractKinveyClient::class.java.getDeclaredField("rootUrl")
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
        assert(field != null)
        field?.isAccessible = true
        try {
            field?.set(client, "https://bmock.kinvey.com/")
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }

    fun cancelMockInvalidConnection() {
        var field: Field? = null
        try {
            field = AbstractKinveyClient::class.java.getDeclaredField("rootUrl")
        } catch (e: NoSuchFieldException) {
            e.printStackTrace()
        }
        assert(field != null)
        field?.isAccessible = true
        try {
            field?.set(client, AbstractClient.DEFAULT_BASE_URL)
        } catch (e: IllegalAccessException) {
            e.printStackTrace()
        }
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testPagedPullBlocking() {
        val store = collection(COLLECTION, Person::class.java, StoreType.CACHE, client)
        client?.syncManager?.clear(COLLECTION)
        testManager?.cleanBackendDataStore(store)

        // Arrange
        val persons = ArrayList<Person>()
        val alvin = Person("Alvin")
        val simon = Person("Simon")
        val theodore = Person("Theodore")
        testManager?.save(store, alvin)
        testManager?.save(store, simon)
        testManager?.save(store, theodore)
        val cacheSizeBefore = testManager?.getCacheSize(StoreType.CACHE, client)
        assertTrue(cacheSizeBefore == 3L)
        client?.cacheManager?.getCache(COLLECTION, Person::class.java, StoreType.CACHE.ttl)?.clear()
        val cacheSizeBetween = testManager?.getCacheSize(StoreType.CACHE, client)
        assertTrue(cacheSizeBetween == 0L)

        // Act
        val response = store.pullBlocking(null, 2)
        assertEquals(3, response.count.toLong())
        assertTrue(response.count.toLong() == testManager?.getCacheSize(StoreType.CACHE, client))
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSyncBlockingPaged() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.syncManager?.clear(COLLECTION)

        var person = Person(TEST_USERNAME)
        testManager?.save(store, person)
        assertTrue(client?.syncManager?.getCount(COLLECTION) == 1L)
        store.pushBlocking()

        person = Person(TEST_USERNAME_2)
        testManager?.save(store, person)
        store.syncBlocking(client?.query(), 1)

        assertTrue(client?.syncManager?.getCount(COLLECTION) == 0L)
        assertTrue(testManager?.find(store, client?.query())?.result?.result?.size == 2)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSyncPaged() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.syncManager?.clear(COLLECTION)

        testManager?.createPersons(store, 10)
        assertTrue(client?.syncManager?.getCount(COLLECTION) == 10L)

        val syncCallback = testManager?.sync(store, null, 1)
        assertNull(syncCallback?.error)
        assertNotNull(syncCallback?.kinveyPushResponse?.successCount)
        assertEquals(10, syncCallback?.kinveyPushResponse?.successCount)
        assertNotNull(syncCallback?.result)
        assertEquals(0, syncCallback?.result?.listOfExceptions?.size)
        assertEquals(10, syncCallback?.result?.count)

        assertTrue(client?.syncManager?.getCount(COLLECTION) == 0L)
        assertEquals(10, store.find()?.result?.size)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSyncPagedWithQuery() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.syncManager?.clear(COLLECTION)

        testManager?.createPersons(store, 10)
        assertTrue(client?.syncManager?.getCount(COLLECTION) == 10L)

        val syncCallback = testManager?.sync(store, client?.query()?.equals("username", TEST_USERNAME + 0)
                ?.or(client?.query()?.equals("username", TEST_USERNAME + 1)), 1)
        assertNull(syncCallback?.error)

        assertNotNull(syncCallback?.kinveyPushResponse?.successCount)
        assertEquals(10, syncCallback?.kinveyPushResponse?.successCount)
        assertNotNull(syncCallback?.result)
        assertEquals(0, syncCallback?.result?.listOfExceptions?.size)
        assertEquals(2, syncCallback?.result?.count)

        assertTrue(client?.syncManager?.getCount(COLLECTION) == 0L)
        assertEquals(10, store.find()?.result?.size)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSyncWithAutoPagination() {
        syncAutoPaginationAsync(true)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testSyncWithoutAutoPagination() {
        syncAutoPaginationAsync(false)
    }

    @Throws(InterruptedException::class, IOException::class)
    private fun syncAutoPaginationAsync(isAutoPagination: Boolean) {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        testManager?.cleanBackend(store, StoreType.SYNC)
        client?.syncManager?.clear(COLLECTION)

        testManager?.createPersons(store, 10)
        assertTrue(client?.syncManager?.getCount(COLLECTION) == 10L)

        val syncCallback = testManager?.sync(store, null, isAutoPagination)
        assertNull(syncCallback?.error)
        assertNotNull(syncCallback?.kinveyPushResponse?.successCount)
        assertEquals(10, syncCallback?.kinveyPushResponse?.successCount)
        assertNotNull(syncCallback?.result)
        assertEquals(0, syncCallback?.result?.listOfExceptions?.size)
        assertEquals(10, syncCallback?.result?.count)

        assertTrue(client?.syncManager?.getCount(COLLECTION) == 0L)
        assertEquals(10, store.find()?.result?.size)
    }

    @Test
    @Throws(InterruptedException::class)
    fun testPagedPullNotCorrectItem() {
        val store = collection(Person.COLLECTION_WITH_EXCEPTION, Person::class.java, StoreType.SYNC, client)
        val pullCallback = testManager?.pullCustom(store, null, 2)
        assertTrue(pullCallback?.result?.listOfExceptions?.size == 1)
        assertTrue(pullCallback?.result?.count == 4)
        testManager?.cleanBackend(store, StoreType.SYNC)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testPullBlockingWithAutoPagination() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        client?.syncManager?.clear(COLLECTION)
        testManager?.cleanBackend(store, StoreType.SYNC)
        testManager?.createPersons(store, 5)
        testManager?.push(store)

        client?.cacheManager?.getCache(COLLECTION, Person::class.java, StoreType.SYNC.ttl)?.clear()
        assertEquals(5, store.pullBlocking(client?.query(), true)?.count)
        assertEquals(5L, testManager?.getCacheSize(StoreType.SYNC, client))
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testPullBlockingWithoutAutoPagination() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        client?.syncManager?.clear(COLLECTION)
        testManager?.cleanBackend(store, StoreType.SYNC)

        testManager?.createPersons(store, 5)
        testManager?.push(store)
        client?.cacheManager?.getCache(COLLECTION, Person::class.java, StoreType.SYNC.ttl)?.clear()

        assertEquals(5, store.pullBlocking(client?.query(), false).count)
        assertEquals(5L, testManager?.getCacheSize(StoreType.SYNC, client))
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testPullAsyncWithAutoPagination() {
        pullAutoPaginationAsync(true)
    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun testPullAsyncWithoutAutoPagination() {
        pullAutoPaginationAsync(false)
    }

    @Throws(InterruptedException::class, IOException::class)
    private fun pullAutoPaginationAsync(isAutoPagination: Boolean) {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        client?.syncManager?.clear(COLLECTION)
        testManager?.cleanBackend(store, StoreType.SYNC)
        testManager?.createPersons(store, 5)
        testManager?.push(store)
        client?.cacheManager?.getCache(COLLECTION, Person::class.java, StoreType.SYNC.ttl)?.clear()
        assertEquals(5, testManager?.pullCustom(store, client?.query(), isAutoPagination)?.result?.count)
        assertEquals(5L, testManager?.getCacheSize(StoreType.SYNC, client))
        client?.cacheManager?.getCache(COLLECTION, Person::class.java, StoreType.SYNC.ttl)?.clear()
        assertEquals(5, testManager?.pullCustom(store, null, isAutoPagination)?.result?.count)
        assertEquals(5L, testManager?.getCacheSize(StoreType.SYNC, client))
    }

    @Test
    fun testPagedPullPrecondition() {
        val store = collection(COLLECTION, Person::class.java, StoreType.CACHE, client)
        try {
            store.pull(client?.query(), -1, null)
            assertFalse(true)
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message?.contains("pageSize mustn't be less than 0") == true)
        }
    }

    @Test
    fun testPagedPullPreconditionWithoutQuery() {
        val store = collection(COLLECTION, Person::class.java, StoreType.CACHE, client)
        try {
            store.pull(-1, null)
            assertFalse(true)
        } catch (e: IllegalArgumentException) {
            assertTrue(e.message?.contains("pageSize mustn't be less than 0") == true)
        }
    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun testDeletingItemsFromTheCacheIfItemsWereDeletedAtTheBackendUsingAP() {
        val store = collection(COLLECTION, Person::class.java, StoreType.SYNC, client)
        val networkStore = collection(COLLECTION, Person::class.java, StoreType.NETWORK, client)
        client?.syncManager?.clear(COLLECTION)
        testManager?.cleanBackend(networkStore, StoreType.NETWORK)

        (0..29).forEach { i ->
            testManager?.save(networkStore, Person(TEST_USERNAME_2))
        }
        (0..4).forEach { i ->
            testManager?.save(networkStore, Person(TEST_USERNAME))
        }
        assertEquals(35, testManager?.pullCustom(store, client?.query(), 10)?.result?.count)
        assertEquals(35, store.count())

        testManager?.delete(networkStore, client?.query()?.equals("username", TEST_USERNAME_2))
        assertEquals(5, testManager?.pullCustom(store, client?.query(), 10)?.result?.count)
        assertEquals(5, store.count())
    }

    companion object {
        const val COLLECTION = "PersonsNew"
    }
}